modules/identity-shiro/src/main/java/org/opensearch/identity/shiro/realm/InternalRealm.java

```diff
@@ -0,0 +1,101 @@
+import java.util.Map;
+
+/**
+ * Internal Realm is a custom realm using the internal OpenSearch IdP
+ *
+ * @opensearch.experimental
+ */
+public class InternalRealm extends AuthenticatingRealm {
+    private static final String DEFAULT_REALM_NAME = "internal";
+
+    public static final InternalRealm INSTANCE = new InternalRealm.Builder(DEFAULT_REALM_NAME).build();
+
+    private String realmName;
+
+    private ConcurrentMap<String, User> internalUsers;
+
+    private InternalRealm(String realmName, ConcurrentMap<String, User> internalUsers) {
+        super(new BCryptPasswordMatcher());
+        this.realmName = realmName;
+        this.internalUsers = internalUsers;
+        setAuthenticationTokenClass(UsernamePasswordToken.class);
+    }
+
+    public static final class Builder {
+        private final String name;
+
+        public Builder(String name) {
+            this.name = Objects.requireNonNull(name);
+        }
+
+        public InternalRealm build() {
+            final User adminUser = new User();
+            adminUser.setUsername(new StringPrincipal("admin"));
+            adminUser.setBcryptHash("$2a$12$VcCDgh2NDk07JGN0rjGbM.Ad41qVR/YFJcgHp0UGns5JDymv..TOG"); // Password 'admin'
+            ConcurrentMap<String, User> internalUsers = new ConcurrentHashMap<>(Map.of("admin", adminUser));
+            return new InternalRealm(name, internalUsers);
+        }
+    }
+
+    public User getInternalUser(String principalIdentifier) throws UnknownAccountException {
+        User userRecord = internalUsers.get(principalIdentifier);
+        if (userRecord == null) {
+            throw new UnknownAccountException();
+        }
+        return userRecord;
+    }
+
+    @Override
+    protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken token) throws AuthenticationException {
+        if (token instanceof UsernamePasswordToken) {
+            String username = ((UsernamePasswordToken) token).getUsername();
+            // Look up the user by the provide username
+            User userRecord = getInternalUser(username);
+            // Check for other things, like a locked account, expired password, etc.
+
+            // Verify the user
+            SimpleAuthenticationInfo sai = new SimpleAuthenticationInfo(userRecord.getUsername(), userRecord.getBcryptHash(), realmName);
+            boolean successfulAuthentication = getCredentialsMatcher().doCredentialsMatch(token, sai);
+
+            if (successfulAuthentication) {
+                // Check for anything else that might prevent login (expired password, locked account, etc.)
+                // if (other problems) {
+                // throw new CredentialsException(); // Or something more specific
+                // }
+                // Success!
+                return sai;
+            } else {
+                // Bad password
+                throw new IncorrectCredentialsException();
+            }
+        }
+        // Don't know what to do with this token
```

Comment:
Should we be including this into main? It seems like this section is still a work in progress.